### PR TITLE
feat(dam): Metrics tag store

### PR DESCRIPTION
**Pull request by shruthilayaj**
Store metrics tags in reflux store instead of
making an API request every time to retrieve
the data. 

Similar PR coming up for metrics meta fields.

### CODE REVIEW

File: static/app/types/metrics.tsx [hunk 1]
```diff
@@ -12,6 +12,8 @@ export type MetricsApiResponse = {
   start: string;
 };
 
+export type MetricTagCollection = Record<string, MetricTag>;
```
Comment (shruthilayaj) [inline static/app/types/metrics.tsx:15]: Using the same structure as TagCollection

File: tests/js/spec/utils/withMetricsTags.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import withMetricsTags from 'sentry/utils/withMetricsTags';
+
+describe('withMetricsTags HoC', function () {
+  beforeEach(() => {
+    MetricsTagStore.reset();
+  });
+
+  afterEach(() => {});
+
+  it('works', async function () {
```
Comment (priscilawebdev) [inline tests/js/spec/utils/withMetricsTags.spec.jsx:13]:
```suggestion
  it('works', function () {
```

File: tests/js/spec/utils/withMetricsTags.spec.jsx [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    jest.spyOn(MetricsTagStore, 'trigger');
+    const MyComponent = ({other, metricsTags}) => {
```
Comment (priscilawebdev) [inline tests/js/spec/utils/withMetricsTags.spec.jsx:15]:
```suggestion
    const MyComponent = ({other, metricsTags = {}}) => {
```

File: tests/js/spec/utils/withMetricsTags.spec.jsx [hunk 4]
```diff
@@ -0,0 +1,48 @@
+import {act, mountWithTheme, screen} from 'sentry-test/reactTestingLibrary';
```
Comment (priscilawebdev) [inline tests/js/spec/utils/withMetricsTags.spec.jsx:1]: can we make this file a typescript file?

File: tests/js/spec/utils/withMetricsTags.spec.jsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+      );
+    };
+
+    const Container = withMetricsTags(MyComponent);
+    mountWithTheme(<Container other="value" />);
+
+    // Should forward props.
+    expect(screen.getByText('value')).toBeInTheDocument();
+
+    act(() => {
```
Comment (priscilawebdev) [inline tests/js/spec/utils/withMetricsTags.spec.jsx:31]: do we need act?

File: tests/js/spec/stores/metricsTagStore.spec.jsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import MetricsTagStore from 'sentry/stores/metricsTagStore';
+
+describe('MetricsTagStore', function () {
+  beforeEach(() => {
+    MetricsTagStore.reset();
+  });
+
+  afterEach(() => {});
```
Comment (priscilawebdev) [inline tests/js/spec/stores/metricsTagStore.spec.jsx:8]:
```suggestion
```

File: static/app/views/dashboardsV2/dashboard.tsx [hunk 7]
```diff
@@ -137,10 +138,14 @@ class Dashboard extends Component<Props, State> {
...
   async componentDidMount() {
-    const {isEditing, organization} = this.props;
+    const {isEditing, organization, api} = this.props;
     if (organization.features.includes('dashboard-grid-layout')) {
       window.addEventListener('resize', this.debouncedHandleResize);
     }
+
+    if (organization.features.includes('dashboard-metrics')) {
+      fetchMetricsTags(api, organization.slug);
+    }
```
Comment (matejminar) [inline static/app/views/dashboardsV2/dashboard.tsx:148]: Do we need it both here and in `fetchTags` method?

**Review by matejminar — APPROVED** [review]
Looks good to me.
We will need to think about if we want to autocomplete only tags that make sense for currently selected fields or not.

**Review by matejminar — APPROVED** [review]
Looks good to me.
We will need to think about if we want to autocomplete only tags that make sense for currently selected fields or not.

**Pull request merged**